File: 🟢 Easy/234 - Palindrome Linked List/_234PalindromeLinkedList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* midNode(struct ListNode* head) {
    struct ListNode* slow = head;
    struct ListNode* fast = head;

    while (fast != NULL && fast->next != NULL) {
        slow = slow->next;       // +1
        fast = fast->next->next; // +2
    }

    return slow; // our mid node
}

bool isPalindrome(struct ListNode* head) {
    if (head == NULL || head->next == NULL) {
        return true;
    }

    struct ListNode* prev = NULL;
    struct ListNode* cur = midNode(head);
    struct ListNode* next = NULL;

    while (cur != NULL) {
        next = cur->next;
        cur->next = prev;
        prev = cur;
        cur = next;
    }

    struct ListNode* left = head;
    struct ListNode* right = prev;

    while (right != NULL) {
        if (left->val != right->val) {
            return false;
        }

        right = right->next;
        left = left->next;
    }

    return true;
}

File: 🟢 Easy/234 - Palindrome Linked List/_234PalindromeLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {boolean}
 */
var midNode = function (head) {
    let slow = head;
    let fast = head;

    while (fast != null && fast.next != null) {
        slow = slow.next; // +1
        fast = fast.next.next; // +2 
    }

    return slow; // our mid node
}

var isPalindrome = function (head) {
    if (head == null || head.next == null) {
        return true;
    }

    let prev = null;
    let cur = midNode(head);
    let next = null;

    while (cur != null) {
        next = cur.next;
        cur.next = prev;
        prev = cur;
        cur = next;
    }

    let left = head;
    let right = prev;

    while (right != null) {
        if (left.val != right.val) {
            return false;
        }

        right = right.next;
        left = left.next;
    }

    return true;
};